lib/auth/init_test.go

```diff
@@ -158,85 +158,124 @@ func TestBadIdentity(t *testing.T) {
-		StaticTokens: []services.ProvisionTokenV1{},
-	})
-	require.NoError(t, err)
-
-	ac := InitConfig{
-		DataDir:        tempDir,
-		HostUUID:       "00000000-0000-0000-0000-000000000000",
-		NodeName:       "foo",
-		Backend:        bk,
-		Authority:      testauthority.New(),
-		ClusterConfig:  services.DefaultClusterConfig(),
-		ClusterName:    clusterName,
-		StaticTokens:   staticTokens,
-		AuthPreference: ap,
-	}
-	as, err := Init(ac)
+	conf := setupConfig(t)
+	conf.AuthPreference = ap
+	as, err := Init(conf)
 	require.NoError(t, err)
 	defer as.Close()
 
 	cap, err := as.GetAuthPreference()
 	require.NoError(t, err)
-	require.Equal(t, cap.GetType(), "local")
+	require.Equal(t, cap.GetType(), constants.Local)
 	require.Equal(t, cap.GetSecondFactor(), constants.SecondFactorU2F)
-
 	u, err := cap.GetU2F()
 	require.NoError(t, err)
 	require.Equal(t, u.AppID, "foo")
 	require.Equal(t, u.Facets, []string{"bar", "baz"})
 }
 
-func TestClusterID(t *testing.T) {
-	tempDir := t.TempDir()
+// TestAuthPreferenceResetConfigFileToDefaults ensures that a config-file
+// AuthPreference can be reset back to defaults.
+func TestAuthPreferenceResetConfigFileToDefaults(t *testing.T) {
+	// Simulate a server with auth preference from config file.
+	var err error
+	conf := setupConfig(t)
+	conf.AuthPreference, err = services.NewAuthPreferenceFromConfigFile(services.AuthPreferenceSpecV2{
+		SecondFactor: constants.SecondFactorOff,
+	})
+	require.NoError(t, err)
+	authServer, err := Init(conf)
+	require.NoError(t, err)
+	defer authServer.Close()
 
-	bk, err := lite.New(context.TODO(), backend.Params{"path": tempDir})
+	// Verify the stored auth preference is labelled as originating from config
+	// file.
+	storedAuthPref, err := authServer.GetAuthPreference()
 	require.NoError(t, err)
+	require.True(t, storedAuthPref.IsFromConfigFile())
+	defaultSecondFactor := services.DefaultAuthPreference().GetSecondFactor()
+	require.NotEqual(t, defaultSecondFactor, storedAuthPref.GetSecondFactor())
 
-	clusterName, err := services.NewClusterName(services.ClusterNameSpecV2{
-		ClusterName: "me.localhost",
-	})
+	// Reset the auth preference back to default.
+	conf.AuthPreference = services.DefaultAuthPreference()
+	authServer, err = Init(conf)
 	require.NoError(t, err)
+	defer authServer.Close()
 
-	authPreference, err := services.NewAuthPreference(services.AuthPreferenceSpecV2{
-		Type: "local",
+	// Verify the stored auth preference is now labelled as originating from
+	// defaults.
+	storedAuthPref, err = authServer.GetAuthPreference()
+	require.NoError(t, err)
+	require.True(t, storedAuthPref.IsFromDefaults())
+	require.Equal(t, defaultSecondFactor, storedAuthPref.GetSecondFactor())
+}
+
+// TestAuthPreferenceResetDynamicToConfigFile ensures that a dynamically
+// configured AuthPreference can be reset back to a config-file preference.
+func TestAuthPreferenceResetDynamicToConfigFile(t *testing.T) {
```

Comment:
add a test for the opposite - attempting to set dynamic preference when a static one is set

Reply:
Ping on this